[CSS] Implement supports() after @import rule

mdubet opened the pull request:
#### 8338db44d91898abef308bf185572c5fd0fdfc69
<pre>
[CSS] Implement supports() after @import rule
<a href="https://bugs.webkit.org/show_bug.cgi?id=256180">https://bugs.webkit.org/show_bug.cgi?id=256180</a>
<a href="https://rdar.apple.com/109060734">rdar://109060734</a>

Reviewed by Antti Koivisto.

<a href="https://drafts.csswg.org/css-cascade-5/#conditional-import">https://drafts.csswg.org/css-cascade-5/#conditional-import</a>
<a href="https://drafts.csswg.org/css-conditional-3/#typedef-supports-condition">https://drafts.csswg.org/css-conditional-3/#typedef-supports-condition</a>

The syntax is: [ supports( [ &lt;supports-condition&gt; | &lt;declaration&gt; ] ) ]?

Like for the CSS.window.supports API, the supports() after @import
allows bare declaration.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/import-conditions-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/import-conditions.html:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssimportrule-expected.txt:
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/css/css-cascade/import-conditions-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/import-conditions-expected.txt: Added.
* Source/WebCore/css/CSSImportRule.cpp:
(WebCore::CSSImportRule::layerName const):
(WebCore::CSSImportRule::supportsText const):
(WebCore::CSSImportRule::cssTextInternal const):
* Source/WebCore/css/CSSImportRule.h:
* Source/WebCore/css/CSSImportRule.idl:
* Source/WebCore/css/StyleRuleImport.cpp:
(WebCore::StyleRuleImport::create):
(WebCore::StyleRuleImport::StyleRuleImport):
* Source/WebCore/css/StyleRuleImport.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSupportsCondition):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeImportRule):
(WebCore::CSSParserImpl::consumeSupportsRule):
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
(WebCore::CSSSupportsParser::supportsCondition):
* Source/WebCore/css/parser/CSSSupportsParser.h:
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addRulesFromSheetContents):

Canonical link: <a href="https://commits.webkit.org/273591@main">https://commits.webkit.org/273591@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4405df0afde219c275e7a073373bd8a7f1e952d5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35903 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14847 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38096 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38625 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32305 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37122 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17266 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11873 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36457 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12552 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31901 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11023 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/36089 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11039 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32069 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39874 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32621 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32394 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36989 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-cascade/import-conditions.html (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11210 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9114 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35066 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/35880 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12959 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/31725 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/42455 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8184 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11721 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12040 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -589,10 +589,29 @@ RefPtr<StyleRuleImport> CSSParserImpl::consumeImportRule(CSSParserTokenRange pre
 
+    struct Supports {
+        String text;
+        bool supported { true };
```
`conditionMatches` or similar?

anttijk commented on Source/WebCore/css/StyleRuleImport.h:
```diff
@@ -77,7 +79,7 @@ class StyleRuleImport final : public StyleRuleBase {
     friend class ImportedStyleSheetClient;
 
-    StyleRuleImport(const String& href, MQ::MediaQueryList&&, std::optional<CascadeLayerName>&&);
+    StyleRuleImport(const String& href, MQ::MediaQueryList&&, std::optional<CascadeLayerName>&&, String&& supports, bool supportsSupported);
```
Can you pass both these as a struct (like `SupportsCondition`)? You already have `struct Supports` in the caller. Then you need fewer awkward names and no magical booleans.

anttijk commented on Source/WebCore/css/StyleRuleImport.h:
```diff
@@ -87,6 +89,8 @@ class StyleRuleImport final : public StyleRuleBase {
     std::optional<CascadeLayerName> m_cascadeLayerName;
     CachedResourceHandle<CachedCSSStyleSheet> m_cachedSheet;
+    String m_supports;
+    bool m_supportsSupported { true };
```
also use the struct here.

mdubet committed:
[CSS] Implement supports() after @import rule
https://bugs.webkit.org/show_bug.cgi?id=256180
rdar://109060734

Reviewed by Antti Koivisto.

https://drafts.csswg.org/css-cascade-5/#conditional-import
https://drafts.csswg.org/css-conditional-3/#typedef-supports-condition

The syntax is: [ supports( [ <supports-condition> | <declaration> ] ) ]?

Like for the CSS.window.supports API, the supports() after @import
allows bare declaration.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/import-conditions-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/import-conditions.html:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssimportrule-expected.txt:
* LayoutTests/platform/gtk-wk2/imported/w3c/web-platform-tests/css/css-cascade/import-conditions-expected.txt: Added.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/import-conditions-expected.txt: Added.
* Source/WebCore/css/CSSImportRule.cpp:
(WebCore::CSSImportRule::layerName const):
(WebCore::CSSImportRule::supportsText const):
(WebCore::CSSImportRule::cssTextInternal const):
* Source/WebCore/css/CSSImportRule.h:
* Source/WebCore/css/CSSImportRule.idl:
* Source/WebCore/css/StyleRuleImport.cpp:
(WebCore::StyleRuleImport::create):
(WebCore::StyleRuleImport::StyleRuleImport):
* Source/WebCore/css/StyleRuleImport.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSupportsCondition):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeImportRule):
(WebCore::CSSParserImpl::consumeSupportsRule):
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
(WebCore::CSSSupportsParser::supportsCondition):
* Source/WebCore/css/parser/CSSSupportsParser.h:
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addRulesFromSheetContents):

Canonical link: https://commits.webkit.org/273591@main
